components/service/glean/src/main/java/mozilla/components/service/glean/firstrun/FileFirstRunDetector.kt

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.firstrun
+
+import android.support.annotation.VisibleForTesting
+import mozilla.components.support.base.log.logger.Logger
+import java.io.File
+import java.io.IOException
+
+/**
+ * A [FirstRunDetector] implementation that uses the presence
+ * of a specific file in the User's app directory to determine
+ * if this is a first run or not.
+ *
+ * @param gleanDataDir the directory, within the application's data dir, that
+ *        contains glean files.
+ */
+internal class FileFirstRunDetector(
+    private val gleanDataDir: String
+) : FirstRunDetector {
+    private val logger = Logger("glean/FileFirstRunDetector")
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    companion object {
+        const val FIRST_RUN_FILENAME = "glean_already_ran"
+    }
+
+    private val firstRun: Boolean by lazy {
+        !checkFirstRunFileExists()
+    }
+
+    /**
+     * Whether or not the first run file exists in the glean's data directory.
+     *
+     * @return true if the file exists (regardless if that's a dir or a file), false
+     *         if it doesn't exist or an exception was thrown.
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun checkFirstRunFileExists(): Boolean {
+        val f = File(gleanDataDir, FIRST_RUN_FILENAME)
+        return try {
+            f.exists()
+        } catch (e: SecurityException) {
+            logger.error("There was a problem testing for ${f.absolutePath}")
+            false
+        }
```

Comment:
Why are you catching `SecurityException` here? Are you expecting this file to (sometimes?) live in the "public" storage (requiring a permission) instead of the app local one? If this is always stored in the app directory then I don't think we should see a SecurityException.

Reply:
Ah, no. This should only live in the app directory :) Thanks for pointing this out!